batteries/difftext/difftext.luau

```diff
@@ -0,0 +1,193 @@
+	seen[0] = {}
+	seen[0][0] = true
+
+	while editGraphDeque.len > 0 do
+		local curNode = editGraphDeque:popfront()
+
+		if not seen[curNode.x] then
+			seen[curNode.x] = {}
+		end
+		if not seen[curNode.x + 1] then
+			seen[curNode.x + 1] = {}
+		end
+
+		-- we've reached bottom right vertex; indicates full edit from a -> b
+		if curNode.x == #a and curNode.y == #b then
+			-- reconstruct diff operations by tracing path from termination node to start
+			while curNode.prev do
+				local diffOperation = edgeToDiff(curNode, a, b)
+				table.insert(diff, 1, diffOperation)
+				curNode = curNode.prev
+			end
+			break
+		end
+
+		-- add neighbors (diag prioritzed, then deletion, then insertion)
+		if hasDiagonalEdge(curNode, a, b) then
+			if not seen[curNode.x + 1][curNode.y + 1] then
+				seen[curNode.x + 1][curNode.y + 1] = true
+				editGraphDeque:pushfront(editGraphNode(curNode.x + 1, curNode.y + 1, curNode))
+			end
+			continue
+		end
+
+		if curNode.x < #a then
+			if not seen[curNode.x + 1][curNode.y] then
+				seen[curNode.x + 1][curNode.y] = true
+				editGraphDeque:pushback(editGraphNode(curNode.x + 1, curNode.y, curNode))
+			end
+		end
+
+		if curNode.y < #b then
+			if not seen[curNode.x][curNode.y + 1] then
+				seen[curNode.x][curNode.y + 1] = true
+				editGraphDeque:pushback(editGraphNode(curNode.x, curNode.y + 1, curNode))
+			end
+		end
+	end
+
+	return diff
+end
+
+local function myersdiffbyline(a: string, b: string)
+	return myersdiff(a:split("\n"), b:split("\n"))
+end
+
+local function myersdiffbychar(a: string, b: string)
+	return myersdiff(a, b)
+end
+
+local function applyDiff(
+	diff: Diff,
+	a: string,
+	options: {
+		byLine: boolean?,
+	}?
+)
+	local separator = if options and options.byLine then "\n" else ""
+	local src = a:split(separator)
+
+	local result = {}
+	local srcCursor = 1 -- Points to the current char in src string
+	for i, op in diff do
+		if op.key == "EQUAL" then
+			local actual = src[srcCursor]
+			if actual ~= op.text then
+				error(
+					string.format(
+						"Patch Error at op %d: Expected EQUAL '%s' but source had '%s'",
+						i,
+						op.text,
+						tostring(actual)
+					)
+				)
+			end
+
+			table.insert(result, op.text)
+			srcCursor = srcCursor + 1
+		elseif op.key == "DELETE" then
+			local actual = src[srcCursor]
+			if actual ~= op.text then
+				error(
+					string.format(
+						"Patch Error at op %d: Expected DELETE '%s' but source had '%s'",
+						i,
+						op.text,
+						tostring(actual)
+					)
+				)
+			end
+
+			srcCursor = srcCursor + 1
+		elseif op.key == "ADD" then
+			table.insert(result, op.text)
+		end
+	end
+
+	return table.concat(result, separator)
+end
+
+local function diff(
+	a: string,
+	b: string,
+	options: {
+		byLine: boolean?,
+	}?
+)
+	if options and options.byLine then
+		return myersdiffbyline(a, b)
+	else
+		return myersdiffbychar(a, b)
+	end
```

Comment:
my personal preference would be to diff by line as default, but maybe bike shed this in proj-lute